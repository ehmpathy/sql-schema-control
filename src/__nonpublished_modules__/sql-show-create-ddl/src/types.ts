export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
