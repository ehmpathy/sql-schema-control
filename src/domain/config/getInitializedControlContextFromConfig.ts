/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
