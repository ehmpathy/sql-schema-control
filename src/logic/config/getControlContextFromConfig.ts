import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
