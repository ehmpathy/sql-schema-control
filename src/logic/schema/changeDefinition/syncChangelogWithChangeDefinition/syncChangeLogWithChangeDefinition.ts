import { mysql as prepareMySQL, pg as preparePgSQL } from 'yesql';

import { DatabaseConnection, ChangeDefinition, DatabaseLanguage } from '../../../../types';

export const syncChangeLogWithChangeDefinition = async ({
  connection,
  definition,
}: {
  connection: DatabaseConnection;
  definition: ChangeDefinition;
}) => {
  // mysql support
  if (connection.language === DatabaseLanguage.MYSQL) {
    return await connection.query(
      prepareMySQL(`
INSERT INTO schema_control_change_log
  (change_id, change_hash, change_content)
VALUES
  (:change_id, :change_hash, :change_content)
ON DUPLICATE KEY UPDATE
  change_hash = :change_hash,
  change_content = :change_content;
      `)({
        change_id: definition.id,
        change_hash: definition.hash,
        change_content: definition.sql,
      }),
    );
  }

  // postgres support
  if (connection.language === DatabaseLanguage.POSTGRES) {
    const prepared = preparePgSQL(`
INSERT INTO schema_control_change_log
  (change_id, change_hash, change_content)
VALUES
  (:change_id, :change_hash, :change_content)
ON CONFLICT (change_id) DO UPDATE
SET
  updated_at = now(),
  change_hash = excluded.change_hash,
  change_content = excluded.change_content;
      `)({
      change_id: definition.id,
      change_hash: definition.hash,
      change_content: definition.sql,
    });
    return await connection.query({ sql: prepared.text, values: prepared.values });
  }

  // if none of the above, throw error
  throw new Error('unsupported database language');
};
