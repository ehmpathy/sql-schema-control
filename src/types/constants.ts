export enum DefinitionType {
  RESOURCE = 'resource',
  CHANGE = 'change',
}

export enum DatabaseLanguage {
  MYSQL = 'mysql',
}

export interface DatabaseConnection {
  query: (args: { sql: string, values?: any[] }) => Promise<any>;
  end: () => Promise<void>;
}

export enum ChangeDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  UP_TO_DATE = 'UP_TO_DATE',
  OUT_OF_DATE = 'OUT_OF_DATE',
}
