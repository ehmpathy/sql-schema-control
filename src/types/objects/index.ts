export { ChangeDefinition } from './ChangeDefinition';
export { ConnectionConfig } from './ConnectionConfig';
export { ControlConfig } from './ControlConfig';
export { ControlContext } from './ControlContext';
