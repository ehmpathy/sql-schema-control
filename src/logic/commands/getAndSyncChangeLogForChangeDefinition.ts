import chalk from 'chalk';

import { ChangeDefinition } from '../../types';
import { getControlContextFromConfig } from '../config/getControlContextFromConfig';
import { syncChangeLogWithChangeDefinition } from '../schema/changeDefinition/syncChangelogWithChangeDefinition';
import { getReferenceIdForDefinition } from '../schema/getReferenceIdForDefinition';
import { getColoredActionTitle } from './utils/getColoredActionTitle';

export const getAndSyncChangeLogForChangeDefinition = async ({
  configPath,
  changeId,
}: {
  configPath: string;
  changeId: string;
}) => {
  // 1. get the control context
  const context = await getControlContextFromConfig({ configPath });

  // 2. find the change definition with this id
  const targetDefinition = context.definitions.find(
    (definition) => getReferenceIdForDefinition({ definition }) === `change:${changeId}`,
  );
  if (!targetDefinition) throw new Error(`could not find a definition with referenceId 'change:${changeId}'`);
  if (!(targetDefinition instanceof ChangeDefinition)) throw new Error('can only sync state for change definitions');

  // 3. sync change definition state manually
  await syncChangeLogWithChangeDefinition({ connection: context.connection, definition: targetDefinition });

  // 4. close the connection
  await context.connection.end();

  // 5. notify the user it was successful
  const successMessage = `  ${chalk.green('âœ”')} ${getColoredActionTitle({
    actionChalk: chalk.green,
    action: 'SYNC',
    definition: targetDefinition,
  })}`;
  console.log('\n', successMessage); // tslint:disable-line no-console
};
