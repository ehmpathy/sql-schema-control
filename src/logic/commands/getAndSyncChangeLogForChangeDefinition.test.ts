import sha256 from 'simple-sha256';

import { getControlContextFromConfig } from '../config/getControlContextFromConfig';
import { getAndSyncChangeLogForChangeDefinition } from './getAndSyncChangeLogForChangeDefinition';
import { ChangeDefinition } from '../../types';
import { syncChangeLogWithChangeDefinition } from '../schema/changeDefinition/syncChangelogWithChangeDefinition';
import { stdout } from 'stdout-stderr';
import chalk from 'chalk';

jest.mock('../config/getControlContextFromConfig');
const getControlContextFromConfigMock = getControlContextFromConfig as jest.Mock;
const exampleContext = {
  connection: {
    end: jest.fn(),
  },
  definitions: [
    new ChangeDefinition({
      id: '__ID__',
      path: '__PATH__',
      sql: '__SQL__',
      hash: sha256.sync('__SQL__'),
    }),
  ],
};
getControlContextFromConfigMock.mockResolvedValue(exampleContext);

jest.mock('../schema/changeDefinition/syncChangelogWithChangeDefinition');
const syncChangeLogWithChangeDefinitionMock = syncChangeLogWithChangeDefinition as jest.Mock;

describe('getAndDisplayPlans', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should get the control context from config', async () => {
    await getAndSyncChangeLogForChangeDefinition({ configPath: '__CONFIG_PATH__', changeId: '__ID__' });
    expect(getControlContextFromConfigMock.mock.calls.length).toEqual(1);
    expect(getControlContextFromConfigMock.mock.calls[0][0]).toEqual({
      configPath: '__CONFIG_PATH__',
    });
  });
  it('should throw an error if no definition with the changeId is found', async () => {
    try {
      await getAndSyncChangeLogForChangeDefinition({
        configPath: '__CONFIG_PATH__',
        changeId: 'change:__DIFFERENT_ID__',
      });
      throw new Error('should not reach here');
    } catch (error) {
      expect(error.message).toContain('could not find a definition with referenceId');
    }
  });
  it('should syncChangeLogWithChangeDefinitionMock', async () => {
    await getAndSyncChangeLogForChangeDefinition({ configPath: '__CONFIG_PATH__', changeId: '__ID__' });
    expect(syncChangeLogWithChangeDefinitionMock).toHaveBeenCalledTimes(1);
    expect(syncChangeLogWithChangeDefinitionMock).toHaveBeenCalledWith(
      expect.objectContaining({
        definition: exampleContext.definitions[0],
      }),
    );
  });
  it('should close the connection', async () => {
    await getAndSyncChangeLogForChangeDefinition({ configPath: '__CONFIG_PATH__', changeId: '__ID__' });
    expect(exampleContext.connection.end.mock.calls.length).toEqual(1);
  });
  it('should log that the change log was synced successfully', async () => {
    // setup env to monitor output
    stdout.stripColor = false; // dont strip color
    stdout.start();

    // run it
    await getAndSyncChangeLogForChangeDefinition({ configPath: '__CONFIG_PATH__', changeId: '__ID__' });

    // check that output was expected
    stdout.stop();
    const output = stdout.output
      .split('\n')
      .filter((line) => !line.includes('console.log'))
      .join('\n'); // strip the console log portion
    expect(output).toContain(chalk.green('[SYNC]'));
    expect(output).toMatchSnapshot();
  });
});
