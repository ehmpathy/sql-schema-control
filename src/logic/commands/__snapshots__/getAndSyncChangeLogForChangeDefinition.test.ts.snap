// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAndDisplayPlans should log that the change log was synced successfully 1`] = `
"    
       [32mâœ”[39m [1m[32m[SYNC][39m __PATH__ [90m(change:__ID__)[39m[22m
[2m[22m
[2m      [2mat [22m[2msrc/logic/commands/getAndSyncChangeLogForChangeDefinition.ts[2m:45:11[22m[2m[22m

"
`;
